### PR TITLE
Rework query highlight/crop parameters

### LINKED ISSUE #549: Query parameter to return only _formatted object
I'm indexing a lot of large text documents. Each can easily be 500-800 KB of text, which I put in a field called full_text. When searching I specify `attributesToCrop=full_text&cropLength=50` and get the result I want in the _formatted object. However, for each hit, I also get the entire full_text back, which is wasteful; 10 hits can result in 5 MB of JSON, whereas it would've been ~40 KB or so without the whole full_text. I can't use `attributesToRetrieve`, because I *do* want the full_text field -- but only the cropped version.

So, it would be great to have the option "just give me the _formatted object and not the non-formatted".

### PR DESCRIPTION
**Pull request by qdequele**
Fixes #549

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,3 +7,4 @@
   - Fixed the core benchmark (#576)
   - Publish an ARMv7 and ARMv8 binaries on releases (#540 and #581)
   - Fixing a bug where the result of the update status after the first update was empty (#542)
+  - Improve search highlight and crop (#549 and #565)
```
Comment (Kerollmops) [inline CHANGELOG.md:10]: Please fix #598 in this PR and add it to the changelog.

File: meilisearch-http/src/helpers/meilisearch.rs [hunk 2]
```diff
@@ -224,36 +224,49 @@ impl<'a> SearchBuilder<'a> {
...
-            // retrieve the content of document in kv store
-            let mut fields: Option<HashSet<&str>> = None;
-            if let Some(attributes_to_retrieve) = &self.attributes_to_retrieve {
-                let mut set = HashSet::new();
-                for field in attributes_to_retrieve {
-                    set.insert(field.as_str());
+        let mut all_attributes: HashSet<&str> = HashSet::new();
+        let mut all_formatted: HashSet<&str> = HashSet::new();
+
+        match &self.attributes_to_retrieve {
```
[the remark this answers is not in the document]
Comment (qdequele) [reply meilisearch-http/src/helpers/meilisearch.rs:235]: I handle it here https://github.com/meilisearch/MeiliSearch/pull/580/files#diff-0745be2d07d612c5cc549ec961696b6dR62

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,3 +8,4 @@
   - Publish an ARMv7 and ARMv8 binaries on releases (#540 and #581)
   - Fixed a bug where the result of the update status after the first update was empty (#542)
   - Fixed a bug where stop words were not handled correctly (#594)
+  - Improve search add wildcard on attributes to retrieve, highlight and crop (#549, #565 and #598)
```
Comment (Kerollmops) [inline CHANGELOG.md:11]:
```suggestion
  - Support wildcard on attributes to retrieve, highlight, and crop (#549, #565, and #598)
```

File: meilisearch-http/src/helpers/meilisearch.rs [hunk 4]
```diff
@@ -224,36 +224,51 @@ impl<'a> SearchBuilder<'a> {
...
+                }
+
+                all_attributes.extend(&all_formatted);
+            },
+            None => {
+                all_attributes.extend(schema.displayed_name());
+                // If we specified at least one attribute to highlight or crop then
+                // all available attributes will be returned in the _formatted field.
+                if self.attributes_to_highlight.is_some() || self.attributes_to_crop.is_some() {
+                    all_formatted.extend(all_attributes.clone());
```
Comment (Kerollmops) [inline meilisearch-http/src/helpers/meilisearch.rs:254]:
```suggestion
                    all_formatted.extend(all_attributes.iter().cloned());
```

File: meilisearch-http/src/helpers/meilisearch.rs [hunk 5]
```diff
@@ -262,13 +277,26 @@ impl<'a> SearchBuilder<'a> {
...
+                let matches =
+                    calculate_matches(matches, self.attributes_to_retrieve.clone(), &schema);
+                Some(matches)
+            } else {
+                None
+            };
+
+            if let Some(attributes_to_retrieve) = &self.attributes_to_retrieve {
+                document.retain(|key, _| attributes_to_retrieve.contains(&key.to_string()))
+            };
```
Comment (Kerollmops) [inline meilisearch-http/src/helpers/meilisearch.rs:299]:
```suggestion
            }
```

File: meilisearch-http/src/helpers/meilisearch.rs [hunk 6]
```diff
@@ -262,13 +277,26 @@ impl<'a> SearchBuilder<'a> {
...
+                    &schema,
+                );
                 formatted = calculate_highlights(&formatted, &matches, attributes_to_highlight);
             }
 
-            let matches_info = if self.matches { Some(matches) } else { None };
+            let matches_info = if self.matches {
+                let matches =
+                    calculate_matches(matches, self.attributes_to_retrieve.clone(), &schema);
+                Some(matches)
```
Comment (Kerollmops) [inline meilisearch-http/src/helpers/meilisearch.rs:292]:
```suggestion
                Some(calculate_matches(matches, self.attributes_to_retrieve.clone(), &schema))
```

File: meilisearch-http/src/routes/search.rs [hunk 7]
```diff
@@ -53,45 +54,75 @@ pub async fn search_with_url_query(ctx: Request<Data>) -> SResult<Response> {
...
     }
 
-    if let Some(attributes_to_retrieve) = query.attributes_to_retrieve {
-        for attr in attributes_to_retrieve.split(',') {
-            search_builder.add_retrievable_field(attr.to_string());
+    let available_attributes = schema.displayed_name();
+    let mut restricted_attributes: HashSet<&str>;
+    match &query.attributes_to_retrieve {
+        Some(attributes_to_retrieve) => {
+            let attributes_to_retrieve: HashSet<&str> = attributes_to_retrieve.split(',').collect();
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/search.rs:61]:
```suggestion
            let attributes_to_retrieve = attributes_to_retrieve.split(',').collect();
```

File: meilisearch-http/src/routes/search.rs [hunk 8]
```diff
@@ -53,45 +54,75 @@ pub async fn search_with_url_query(ctx: Request<Data>) -> SResult<Response> {
...
-        } else {
-            let attributes_to_crop = attributes_to_crop
-                .split(',')
-                .map(|r| (r.to_string(), crop_length))
-                .collect();
-            search_builder.attributes_to_crop(attributes_to_crop);
+        let default_length: usize = query.crop_length.unwrap_or(200);
+        let mut final_attributes: HashMap<String, usize> = HashMap::new();
+
+        for attribute in attributes_to_crop.split(',').collect::<Vec<&str>>().iter() {
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/search.rs:85]:
```suggestion
        for attribute in attributes_to_crop.split(',') {
```

File: meilisearch-http/src/routes/search.rs (hunk 7 above)
Comment (qdequele) [reply meilisearch-http/src/routes/search.rs:61]: I must give to attributes_to_retrieve a type at that point.

Comment (Kerollmops) [reply meilisearch-http/src/routes/search.rs:61]: Ho, yes, my mistake, I thought it was `restricted_attributes` sorry.

File: CHANGELOG.md (hunk 3 above)
Comment (Kerollmops) [reply CHANGELOG.md]: Your forget the comma between "highlight" and "crop"...

File: meilisearch-http/src/routes/search.rs [hunk 9]
```diff
@@ -53,45 +54,75 @@ pub async fn search_with_url_query(ctx: Request<Data>) -> SResult<Response> {
...
-                .map(|attr| (attr.to_string(), crop_length))
-                .collect();
-            search_builder.attributes_to_crop(attributes_to_crop);
-        } else {
-            let attributes_to_crop = attributes_to_crop
-                .split(',')
-                .map(|r| (r.to_string(), crop_length))
-                .collect();
-            search_builder.attributes_to_crop(attributes_to_crop);
+        let default_length: usize = query.crop_length.unwrap_or(200);
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/search.rs:82]:
```suggestion
        let default_length = query.crop_length.unwrap_or(200);
```

Rust already know the type of `query.crop_length`.

File: meilisearch-http/src/routes/search.rs [hunk 10]
```diff
@@ -53,45 +54,75 @@ pub async fn search_with_url_query(ctx: Request<Data>) -> SResult<Response> {
...
+                    }
+                },
+                Some(attr) => {
+                    if available_attributes.contains(attr) {
+                        final_attributes.insert(attr.to_string(), length);
+                    } else {
+                        warn!("The attributes {:?} present in attributesToCrop parameter doesn't exist", attr);
+                    }
+                },
+                None => ()
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/search.rs:102]:
```suggestion
                None => (),
```

**Pull request merged**